File: aidldemo/src/main/java/com/example/aidldemo/MainActivity.java
package com.example.aidldemo;

import androidx.appcompat.app.AppCompatActivity;

import android.content.ComponentName;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.IBinder;
import android.os.RemoteException;
import android.util.Log;
import android.view.View;

import com.example.servicedemo.IMyAidlInterface;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {

    private static final String TAG = "MainActivity";
    /**
     * 用于绑定客户端和服务
     */
    ServiceConnection serviceConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            IMyAidlInterface iMyAidlInterface = IMyAidlInterface.Stub.asInterface(service);
            try {
                iMyAidlInterface.showProgress();
            } catch (RemoteException e) {
                e.printStackTrace();
            }
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {

        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        initEvent();
    }

    private void initEvent() {
        findViewById(R.id.start_service).setOnClickListener(this);
        findViewById(R.id.stop_service).setOnClickListener(this);
        findViewById(R.id.bind_service).setOnClickListener(this);
        findViewById(R.id.unbind_service).setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.start_service:
                Intent intent = new Intent();
                intent.setAction("com.example.myservice");
                intent.setPackage("com.example.servicedemo");
                startService(intent);
                break;
            case R.id.stop_service:
                Intent intent1 = new Intent("com.example.myservice");
                intent1.setPackage("com.example.servicedemo");
                stopService(intent1);
                break;
            case R.id.bind_service:
                Intent intent2 = new Intent("com.example.myservice");
                intent2.setPackage("com.example.servicedemo");
                bindService(intent2, serviceConnection, BIND_AUTO_CREATE);
                break;
            case R.id.unbind_service:
                unbindService(serviceConnection);
                break;
        }
    }
}

File: servicedemo/src/main/java/com/jonesyong/servicedemo/MainActivity.kt
package com.jonesyong.servicedemo

import android.app.Service
import android.content.ComponentName
import android.content.Intent
import android.content.ServiceConnection
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.os.IBinder
import android.os.Messenger
import android.util.Log
import android.widget.Button
import android.widget.TextView
import com.jonesyong.servicedemo.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity() {

    companion object {
        private const val TAG = "LocalService"
    }

    private lateinit var binding: ActivityMainBinding
    private lateinit var localService: LocalService
    private lateinit var startService: Button
    private lateinit var stopService: Button
    private lateinit var bindService: Button
    private lateinit var unBindService: Button
    private lateinit var getBindData: Button
    private lateinit var showTextView: TextView
    private lateinit var messengerButton: Button

    private val serviceConnection: ServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName?, service: IBinder?) {
            Log.d(TAG, "Service onServiceConnected")
            val binder: LocalService.LocalBinder = service as LocalService.LocalBinder
            localService = binder.getService()
        }

        override fun onServiceDisconnected(name: ComponentName?) {
            Log.d(TAG, "Service onServiceDisconnected")
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        initView()
        initEvent()
    }

    private fun initEvent() {
        startService.setOnClickListener {
            val intent = Intent(this, LocalService::class.java)
            startService(intent)
        }
        stopService.setOnClickListener {
            val intent = Intent("com.jonesyong.servicedemo.LocalService")
            intent.setPackage(packageName)
            stopService(intent)
        }
        bindService.setOnClickListener {
            val intent = Intent("com.jonesyong.servicedemo.LocalService")
            intent.setPackage(packageName)
            bindService(intent, serviceConnection, Service.BIND_AUTO_CREATE)
        }
        unBindService.setOnClickListener {
            unbindService(serviceConnection)
        }

        // 获取数据、展示数据
        getBindData.setOnClickListener {
            showTextView.text = localService.getCount().toString()
            Log.d(TAG, localService.getCount().toString())
        }

        // 进入启动远程服务的界面
        messengerButton.setOnClickListener {
            startActivity(Intent(this, MessengerActivity::class.java))
        }
    }

    private fun initView() {
        startService = binding.startService
        stopService = binding.stopService
        bindService = binding.bindService
        unBindService = binding.unbindService
        getBindData = binding.obtainText
        showTextView = binding.showText
        messengerButton = binding.remoteMessenger
    }
}
